File: jvm/sargon-android/src/main/java/com/radixdlt/sargon/extensions/LegacyOlympiaAccountAddress.kt
package com.radixdlt.sargon.extensions

import com.radixdlt.sargon.AccountAddress
import com.radixdlt.sargon.AddressFormat
import com.radixdlt.sargon.IdentityAddress
import com.radixdlt.sargon.LegacyOlympiaAccountAddress
import com.radixdlt.sargon.NetworkId
import com.radixdlt.sargon.PublicKey
import com.radixdlt.sargon.identityAddressFormatted
import com.radixdlt.sargon.legacyOlympiaAccountAddressFormatted
import com.radixdlt.sargon.legacyOlympiaAccountAddressIsLegacyOfBabylon
import com.radixdlt.sargon.legacyOlympiaAccountAddressToBabylonAccountAddress
import com.radixdlt.sargon.legacyOlympiaAccountAddressToString
import com.radixdlt.sargon.newLegacyOlympiaAccountAddressFromPublicKey
import com.radixdlt.sargon.newLegacyOlympiaAccountAddressFromString

@Throws(SargonException::class)
fun LegacyOlympiaAccountAddress.Companion.init(validatingAddress: String) =
    newLegacyOlympiaAccountAddressFromString(string = validatingAddress)

fun LegacyOlympiaAccountAddress.Companion.init(publicKey: PublicKey.Secp256k1) =
    newLegacyOlympiaAccountAddressFromPublicKey(publicKey = publicKey.value)

val LegacyOlympiaAccountAddress.string: String
    get() = legacyOlympiaAccountAddressToString(address = this)

val LegacyOlympiaAccountAddress.networkId: NetworkId
    // We do not allow creation of Non-Mainnet Olympia Addresses.
    get() = NetworkId.MAINNET

fun LegacyOlympiaAccountAddress.toBabylonAddress() =
    legacyOlympiaAccountAddressToBabylonAccountAddress(address = this)

fun LegacyOlympiaAccountAddress.isLegacyOfBabylonAddress(babylon: AccountAddress) =
    legacyOlympiaAccountAddressIsLegacyOfBabylon(
        legacyOlympiaAddress = this,
        babylonAccountAddress = babylon
    )

fun LegacyOlympiaAccountAddress.formatted(
    format: AddressFormat = AddressFormat.DEFAULT
): String = legacyOlympiaAccountAddressFormatted(address = this, format = format)

File: crates/sargon-uniffi/src/profile/v100/address/legacy_olympia_account_address.rs
use crate::prelude::*;
use sargon::LegacyOlympiaAccountAddress as InternalLegacyOlympiaAccountAddress;

#[derive(Clone, PartialEq, Eq, Hash, InternalConversion, uniffi::Record)]
pub struct LegacyOlympiaAccountAddress {
    value: Secp256k1PublicKey,
}

impl From<InternalLegacyOlympiaAccountAddress> for LegacyOlympiaAccountAddress {
    fn from(value: InternalLegacyOlympiaAccountAddress) -> Self {
        Self {
            value: value.public_key.into(),
        }
    }
}

impl Into<InternalLegacyOlympiaAccountAddress> for LegacyOlympiaAccountAddress {
    fn into(self) -> InternalLegacyOlympiaAccountAddress {
        InternalLegacyOlympiaAccountAddress {
            public_key: self.value.into(),
        }
    }
}

#[uniffi::export]
pub fn new_legacy_olympia_account_address_sample() -> LegacyOlympiaAccountAddress
{
    InternalLegacyOlympiaAccountAddress::sample().into()
}

#[uniffi::export]
pub fn new_legacy_olympia_account_address_sample_other(
) -> LegacyOlympiaAccountAddress {
    InternalLegacyOlympiaAccountAddress::sample_other().into()
}

#[uniffi::export]
pub fn new_legacy_olympia_account_address_from_public_key(
    public_key: Secp256k1PublicKey,
) -> LegacyOlympiaAccountAddress {
    InternalLegacyOlympiaAccountAddress::from(public_key.into_internal()).into()
}

#[uniffi::export]
pub fn new_legacy_olympia_account_address_from_string(
    string: String,
) -> Result<LegacyOlympiaAccountAddress> {
    InternalLegacyOlympiaAccountAddress::from_str(&string).map_result()
}

#[uniffi::export]
pub fn legacy_olympia_account_address_to_string(
    address: &LegacyOlympiaAccountAddress,
) -> String {
    address.into_internal().to_string()
}

#[uniffi::export]
pub fn legacy_olympia_account_address_formatted(
    address: &LegacyOlympiaAccountAddress,
    format: AddressFormat,
) -> String {
    address.into_internal().formatted(format.into_internal())
}

#[uniffi::export]
pub fn legacy_olympia_account_address_to_babylon_account_address(
    address: LegacyOlympiaAccountAddress,
) -> AccountAddress {
    address.into_internal().to_babylon_account_address().into()
}

#[uniffi::export]
pub fn legacy_olympia_account_address_is_legacy_of_babylon(
    legacy_olympia_address: &LegacyOlympiaAccountAddress,
    babylon_account_address: &AccountAddress,
) -> bool {
    babylon_account_address
        .into_internal()
        .was_migrated_from_legacy_olympia_account_address(
            &legacy_olympia_address.into_internal(),
        )
}
